src/raftstore/coprocessor/region_collection.rs

```diff
@@ -0,0 +1,675 @@
+/// through the `scheduler`.
+#[derive(Clone)]
+struct EventSender {
+    scheduler: Scheduler<RegionCollectionMsg>,
+}
+
+impl Coprocessor for EventSender {}
+
+impl RegionChangeObserver for EventSender {
+    fn on_region_changed(&self, context: &mut ObserverContext, event: RegionChangeEvent) {
+        let region = context.region().clone();
+        let event = match event {
+            RegionChangeEvent::Create => RaftStoreEvent::CreateRegion { region },
+            RegionChangeEvent::Update => RaftStoreEvent::UpdateRegion { region },
+            RegionChangeEvent::Destroy => RaftStoreEvent::DestroyRegion { region },
+        };
+        self.scheduler
+            .schedule(RegionCollectionMsg::RaftStoreEvent(event))
+            .unwrap();
+    }
+}
+
+impl RoleObserver for EventSender {
+    fn on_role_change(&self, context: &mut ObserverContext, role: StateRole) {
+        let region = context.region().clone();
+        let event = RaftStoreEvent::RoleChange { region, role };
+        self.scheduler
+            .schedule(RegionCollectionMsg::RaftStoreEvent(event))
+            .unwrap();
+    }
+}
+
+/// Create an `EventSender` and register it to given coprocessor host.
+fn register_raftstore_event_sender(
+    host: &mut CoprocessorHost,
+    scheduler: Scheduler<RegionCollectionMsg>,
+) {
+    let event_sender = EventSender { scheduler };
+
+    host.registry
+        .register_role_observer(1, box event_sender.clone());
+    host.registry
+        .register_region_change_observer(1, box event_sender.clone());
+}
+
+/// `RegionCollectionWorker` is the underlying runner of `RegionCollection`. It listens on events
+/// sent by the `EventSender` and maintains the collection of all regions. Role of each region
+/// are also tracked.
+struct RegionCollectionWorker {
+    // region_id -> (Region, State)
+    regions: HashMap<u64, (Region, StateRole)>,
+    // region_id -> prefixed end key ('z' + key)
+    region_ranges: BTreeMap<Vec<u8>, u64>,
+}
+
+impl RegionCollectionWorker {
+    fn new() -> Self {
+        Self {
+            regions: HashMap::default(),
+            region_ranges: BTreeMap::default(),
+        }
+    }
+
+    fn handle_create_region(&mut self, region: Region) {
+        if self.regions.get(&region.get_id()).is_some() {
+            warn!(
+                "region_collection: trying to create new region {} but it already exists. \
+                 try to update it.",
+                region.get_id(),
+            );
+            self.handle_update_region(region);
+            return;
+        }
+
+        self.region_ranges
+            .insert(data_end_key(region.get_end_key()), region.get_id());
+        // TODO: Should we set it follower?
+        self.regions
+            .insert(region.get_id(), (region, StateRole::Follower));
+    }
+
+    fn handle_update_region(&mut self, region: Region) {
+        let mut is_new_region = true;
```

Comment:
It was used in an `if` in this function. The content of that if can't be made into `else` clause of L165's `if` because it borrows mut from `self.region`.

Reply:
got it